### ISSUE TITLE
LinkGenerator throws NPE in async code after update to Grails 3 

### ISSUE THREAD

**Issue by demon101** [grails/grails-core#10435]
### Task List

- [X] Steps to reproduce provided
- [X] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [X] Full description of the issue provided (see below)

### Steps to Reproduce

Simple asynchronous code throw NPE during link generation

```
Domain.async.task {
...
def lk = grailsLinkGenerator.link(controller: 'view', action: 'index', params: [name:name], absolute: true)
...
}
```

> java.lang.NullPointerException: null
> 	at org.apache.catalina.connector.Request.getServletContext(Request.java:1618)
> 	at org.apache.catalina.connector.Request.getContextPath(Request.java:1945)
> 	at org.apache.catalina.connector.RequestFacade.getContextPath(RequestFacade.java:783)
> 	at javax.servlet.http.HttpServletRequestWrapper.getContextPath(HttpServletRequestWrapper.java:189)
> 	at javax.servlet.http.HttpServletRequestWrapper.getContextPath(HttpServletRequestWrapper.java:189)
> 	at asset.pipeline.grails.utils.net.HttpServletRequests.getBaseUrlSansScheme(HttpServletRequests.java:139)
> 	at asset.pipeline.grails.utils.net.HttpServletRequests.getBaseUrlWithScheme(HttpServletRequests.java:123)
> 	at asset.pipeline.grails.utils.net.HttpServletRequests.getBaseUrlWithScheme(HttpServletRequests.java:115)
> 	at asset.pipeline.grails.utils.net.HttpServletRequests$getBaseUrlWithScheme.callStatic(Unknown Source)
> 	at asset.pipeline.grails.CachingLinkGenerator.makeKey(CachingLinkGenerator.groovy:52)
> 	at org.grails.web.mapping.CachingLinkGenerator.link(CachingLinkGenerator.java:78)
> 	at org.grails.web.mapping.DefaultLinkGenerator.link(DefaultLinkGenerator.groovy)
> 	at grails.web.mapping.LinkGenerator$link.call(Unknown Source)
> 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
> 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
> 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)

### Expected Behaviour

Link should be generatied. It works in grails 2

### Actual Behaviour

NPE

### Environment Information

- **Grails Version:** 3.1.15
- **JDK Version:** 8 102

**Comment by jeffbrown** [comment]
See the project at https://github.com/jeffbrown/issue10435.

```
package demo

import grails.web.mapping.LinkGenerator

import static grails.async.Promises.waitAll

class DemoController {

    LinkGenerator grailsLinkGenerator

    def index() {
        String link = '<unknown>'
        def task = Person.async.task {
            link = grailsLinkGenerator.link(controller: 'demo', action: 'foo', params: [name: 'Jeff'])
        }

        waitAll(task)

        render "Link: ${link}"
    }

    def foo(String name) {
        render "Name: $name"
    }
}
```

That appears to work.

Please send a PR to that project which demonstrates the problem.

Thanks for the feedback.

**Comment by demon101** [comment]
@jeffbrown  I hope, problem can be solved by #10437 
 
In my app I has something like this

```
package demo

import grails.web.mapping.LinkGenerator

class DemoService {

    def grailsLinkGenerator

    def test() {
        
        def task = Person.async.task {
            sleep(5000)
            def link = grailsLinkGenerator.link(controller: 'demo', action: 'foo', params: [name: 'Jeff'])
            log.error(link)
        }

        return "done"
    }
}
```

When you plan to release version with the #10437 fix?
I want to test it with my huge app. Overwise I will prepare test app for you

**Comment by demon101** [comment]
@jeffbrown can't reproduce. I hope the bug was coused by hotswap of app during development.
Will add additional logs and will check on prod enviroment
